File: src/lexer/lexer_test.go
package lexer

import (
	"testing"
	"vesti/src/token"
)

func testLexing(t *testing.T, input string, expectedLst []token.Token) {
	t.Helper()

	l := New(input)

	for _, expected := range expectedLst {
		got := l.TakeTok()

		if got.Token.Type != expected.Type {
			t.Fatalf("token types are different in `%v`.\ngot=%X, want=%X",
				got.Span, got.Token.Type, expected.Type)
		}

		if got.Token.Literal != expected.Literal {
			t.Fatalf("token literals are different in `%v`.\ngot=%q, want=%q",
				got.Span, got.Token.Literal, expected.Literal)
		}
	}
}

func TestLexingSymbols(t *testing.T) {
	input := "+-/*!?@&^;_:\".,`|'~"
	expected := []token.Token{
		token.New(token.Plus, "+"),
		token.New(token.Minus, "-"),
		token.New(token.Slash, "/"),
		token.New(token.Star, "*"),
		token.New(token.Bang, "!"),
		token.New(token.Question, "?"),
		token.New(token.At, "@"),
		token.New(token.Ampersand, "&"),
		token.New(token.Superscript, "^"),
		token.New(token.Semicolon, ";"),
		token.New(token.Subscript, "_"),
		token.New(token.Colon, ":"),
		token.New(token.Doublequote, "\""),
		token.New(token.Period, "."),
		token.New(token.Comma, ","),
		token.New(token.Quote2, "`"),
		token.New(token.Vert, "|"),
		token.New(token.Quote, "'"),
		token.New(token.Tilde, "~"),
	}

	testLexing(t, input, expected)
}

func TestLexingWhitespaces(t *testing.T) {
	input := "\t  \t\n\r\n \t\r"
	expected := []token.Token{
		token.New(token.Tab, "\t"),
		token.New(token.Space, " "),
		token.New(token.Space, " "),
		token.New(token.Tab, "\t"),
		token.New(token.Newline, "\n"),
		token.New(token.Newline, "\n"),
		token.New(token.Space, " "),
		token.New(token.Tab, "\t"),
		token.New(token.Newline, "\n"),
	}

	testLexing(t, input, expected)
}

func TestLexingComment(t *testing.T) {
	input := `
# This is a comment!
----
#* This is also a comment.
The difference is that multiple commenting is possible.
*#+`
	expected := []token.Token{
		token.New(token.Newline, "\n"),
		token.New(token.Minus, "-"),
		token.New(token.Minus, "-"),
		token.New(token.Minus, "-"),
		token.New(token.Minus, "-"),
		token.New(token.Newline, "\n"),
		token.New(token.Plus, "+"),
	}

	testLexing(t, input, expected)
}

func TestLexingTextRawLatex(t *testing.T) {
	input := "#-\\TeX and \\LaTeX-##-foo 3.14-#"
	expected := []token.Token{
		token.New(token.RawLatex, "\\TeX and \\LaTeX"),
		token.New(token.RawLatex, "foo 3.14"),
	}

	testLexing(t, input, expected)
}

func TestLexingInlineRawLatex(t *testing.T) {
	input := `##-
\begin{center}
  the \TeX
\end{center}
-##`
	expected := []token.Token{
		token.New(token.RawLatex, `
\begin{center}
  the \TeX
\end{center}
`),
	}

	testLexing(t, input, expected)
}

func TestLexingAsciiString(t *testing.T) {
	input := "This is a string!"
	expected := []token.Token{
		token.New(token.MainString, "This"),
		token.New(token.Space, " "),
		token.New(token.MainString, "is"),
		token.New(token.Space, " "),
		token.New(token.MainString, "a"),
		token.New(token.Space, " "),
		token.New(token.MainString, "string"),
		token.New(token.Bang, "!"),
	}

	testLexing(t, input, expected)
}

func TestLexingUnicodeString(t *testing.T) {
	input := "이것은 무엇인가?"
	expected := []token.Token{
		token.New(token.MainString, "이것은"),
		token.New(token.Space, " "),
		token.New(token.MainString, "무엇인가"),
		token.New(token.Question, "?"),
	}

	testLexing(t, input, expected)
}

func TestLexingNumber(t *testing.T) {
	input := "1 32 -8432 3.2 0.3 32.00"
	expected := []token.Token{
		token.New(token.Integer, "1"),
		token.New(token.Space, " "),
		token.New(token.Integer, "32"),
		token.New(token.Space, " "),
		token.New(token.Integer, "-8432"),
		token.New(token.Space, " "),
		token.New(token.Float, "3.2"),
		token.New(token.Space, " "),
		token.New(token.Float, "0.3"),
		token.New(token.Space, " "),
		token.New(token.Float, "32.00"),
	}

	testLexing(t, input, expected)
}

func TestLexingKeywords(t *testing.T) {
	input := "docclass begenv document mtxt import etxt endenv"
	expected := []token.Token{
		token.New(token.Docclass, "docclass"),
		token.New(token.Space, " "),
		token.New(token.Begenv, "begenv"),
		token.New(token.Space, " "),
		token.New(token.Document, "document"),
		token.New(token.Space, " "),
		token.New(token.Mtxt, "mtxt"),
		token.New(token.Space, " "),
		token.New(token.Import, "import"),
		token.New(token.Space, " "),
		token.New(token.Etxt, "etxt"),
		token.New(token.Space, " "),
		token.New(token.Endenv, "endenv"),
	}

	testLexing(t, input, expected)
}

func TestLexingMathDelimiter(t *testing.T) {
	input := "$ $ $$ $$ $ $$ $ \\) \\[ \\] \\( $"
	expected := []token.Token{
		token.New(token.TextMathStart, "$"),
		token.New(token.Space, " "),
		token.New(token.TextMathEnd, "$"),
		token.New(token.Space, " "),
		token.New(token.InlineMathStart, "\\["),
		token.New(token.Space, " "),
		token.New(token.InlineMathEnd, "\\]"),
		token.New(token.Space, " "),
		token.New(token.TextMathStart, "$"),
		token.New(token.Space, " "),
		token.New(token.InlineMathEnd, "\\]"),
		token.New(token.Space, " "),
		token.New(token.TextMathStart, "$"),
		token.New(token.Space, " "),
		token.New(token.TextMathEnd, "$"),
		token.New(token.Space, " "),
		token.New(token.InlineMathStart, "\\["),
		token.New(token.Space, " "),
		token.New(token.InlineMathEnd, "\\]"),
		token.New(token.Space, " "),
		token.New(token.TextMathStart, "$"),
		token.New(token.Space, " "),
		token.New(token.TextMathEnd, "$"),
	}

	testLexing(t, input, expected)
}

func TestLexingLatexFunctions(t *testing.T) {
	input := "\\foo \\bar@hand"
	expected := []token.Token{
		token.New(token.LatexFunction, "foo"),
		token.New(token.Space, " "),
		token.New(token.LatexFunction, "bar@hand"),
	}

	testLexing(t, input, expected)
}

func TestLexingBasicVesti(t *testing.T) {
	input := `docclass coprime (tikz, korean)
import {
    geometry (a4paper, margin = 0.4in),
    amsmath,
}

document

This is a \LaTeX!
$$
    1 + 1 = \sum_{j=1}^\infty f(x),\qquad mtxt foobar etxt
$$
begenv center
    The TeX
endenv`
	expected := []token.Token{
		token.New(token.Docclass, "docclass"),
		token.New(token.Space, " "),
		token.New(token.MainString, "coprime"),
		token.New(token.Space, " "),
		token.New(token.Lparen, "("),
		token.New(token.MainString, "tikz"),
		token.New(token.Comma, ","),
		token.New(token.Space, " "),
		token.New(token.MainString, "korean"),
		token.New(token.Rparen, ")"),
		token.New(token.Newline, "\n"),
		token.New(token.Import, "import"),
		token.New(token.Space, " "),
		token.New(token.Lbrace, "{"),
		token.New(token.Newline, "\n"),
		token.New(token.Space, " "),
		token.New(token.Space, " "),
		token.New(token.Space, " "),
		token.New(token.Space, " "),
		token.New(token.MainString, "geometry"),
		token.New(token.Space, " "),
		token.New(token.Lparen, "("),
		token.New(token.MainString, "a4paper"),
		token.New(token.Comma, ","),
		token.New(token.Space, " "),
		token.New(token.MainString, "margin"),
		token.New(token.Space, " "),
		token.New(token.Equal, "="),
		token.New(token.Space, " "),
		token.New(token.Float, "0.4"),
		token.New(token.MainString, "in"),
		token.New(token.Rparen, ")"),
		token.New(token.Comma, ","),
		token.New(token.Newline, "\n"),
		token.New(token.Space, " "),
		token.New(token.Space, " "),
		token.New(token.Space, " "),
		token.New(token.Space, " "),
		token.New(token.MainString, "amsmath"),
		token.New(token.Comma, ","),
		token.New(token.Newline, "\n"),
		token.New(token.Rbrace, "}"),
		token.New(token.Newline, "\n"),
		token.New(token.Newline, "\n"),
		token.New(token.Document, "document"),
		token.New(token.Newline, "\n"),
		token.New(token.Newline, "\n"),
		token.New(token.MainString, "This"),
		token.New(token.Space, " "),
		token.New(token.MainString, "is"),
		token.New(token.Space, " "),
		token.New(token.MainString, "a"),
		token.New(token.Space, " "),
		token.New(token.LatexFunction, "LaTeX"),
		token.New(token.Bang, "!"),
		token.New(token.Newline, "\n"),
		token.New(token.InlineMathStart, "\\["),
		token.New(token.Newline, "\n"),
		token.New(token.Space, " "),
		token.New(token.Space, " "),
		token.New(token.Space, " "),
		token.New(token.Space, " "),
		token.New(token.Integer, "1"),
		token.New(token.Space, " "),
		token.New(token.Plus, "+"),
		token.New(token.Space, " "),
		token.New(token.Integer, "1"),
		token.New(token.Space, " "),
		token.New(token.Equal, "="),
		token.New(token.Space, " "),
		token.New(token.LatexFunction, "sum"),
		token.New(token.Subscript, "_"),
		token.New(token.Lbrace, "{"),
		token.New(token.MainString, "j"),
		token.New(token.Equal, "="),
		token.New(token.Integer, "1"),
		token.New(token.Rbrace, "}"),
		token.New(token.Superscript, "^"),
		token.New(token.LatexFunction, "infty"),
		token.New(token.Space, " "),
		token.New(token.MainString, "f"),
		token.New(token.Lparen, "("),
		token.New(token.MainString, "x"),
		token.New(token.Rparen, ")"),
		token.New(token.Comma, ","),
		token.New(token.LatexFunction, "qquad"),
		token.New(token.Space, " "),
		token.New(token.Mtxt, "mtxt"),
		token.New(token.Space, " "),
		token.New(token.MainString, "foobar"),
		token.New(token.Space, " "),
		token.New(token.Etxt, "etxt"),
		token.New(token.Newline, "\n"),
		token.New(token.InlineMathEnd, "\\]"),
		token.New(token.Newline, "\n"),
		token.New(token.Begenv, "begenv"),
		token.New(token.Space, " "),
		token.New(token.MainString, "center"),
		token.New(token.Newline, "\n"),
		token.New(token.Space, " "),
		token.New(token.Space, " "),
		token.New(token.Space, " "),
		token.New(token.Space, " "),
		token.New(token.MainString, "The"),
		token.New(token.Space, " "),
		token.New(token.MainString, "TeX"),
		token.New(token.Newline, "\n"),
		token.New(token.Endenv, "endenv"),
	}

	testLexing(t, input, expected)
}

File: src/experimental/lexer/lexer_test.rs
// Copyright (c) 2022 Sungbae Jeong
//
// This software is released under the MIT License.
// https://opensource.org/licenses/MIT

use super::*;

macro_rules! lexing {
    ($lex: ident) => {{
        let mut lexed = Vec::new();
        let mut token;
        loop {
            token = $lex.next();
            match token.toktype {
                TokenType::Eof => break lexed,
                TokenType::Illegal => panic!(
                    "illegal token {:?} was found on {:?}",
                    $lex.chr0, token.span
                ),
                _ => {}
            }
            lexed.push(token);
        }
    }};
}

macro_rules! token {
    ($toktype: expr, $literal: literal) => {
        Token {
            toktype: $toktype,
            literal: $literal.to_string(),
            span: Span::default(),
        }
    };
}

fn check_same(toks1: Vec<Token>, toks2: Vec<Token>) -> bool {
    if toks1.len() != toks2.len() {
        println!("length expected {:?}, got {:?}", toks1.len(), toks2.len());
        return false;
    }

    for (tok1, tok2) in toks1.into_iter().zip(toks2) {
        if tok1.toktype != tok2.toktype {
            println!("token expected {:?}, got {:?}", tok1.toktype, tok2.toktype);
            return false;
        }
        if tok1.literal != tok2.literal {
            println!(
                "literal expected {:?}, got {:?}",
                tok1.literal, tok2.literal
            );
            return false;
        }
    }

    true
}

#[test]
fn test_lexing_single_symbols() {
    let source = "+*;:\"'`|=<?!,.~#";
    let expected = vec![
        token!(TokenType::Plus, "+"),
        token!(TokenType::Star, "*"),
        token!(TokenType::Semicolon, ";"),
        token!(TokenType::Colon, ":"),
        token!(TokenType::DoubleQuote, "\""),
        token!(TokenType::RightQuote, "'"),
        token!(TokenType::LeftQuote, "`"),
        token!(TokenType::Vert, "|"),
        token!(TokenType::Equal, "="),
        token!(TokenType::Less, "<"),
        token!(TokenType::Question, "?"),
        token!(TokenType::Bang, "!"),
        token!(TokenType::Comma, ","),
        token!(TokenType::Period, "."),
        token!(TokenType::Tilde, "~"),
        token!(TokenType::FntParam, "#"),
    ];

    let mut lex = Lexer::new(source);
    let lexed = lexing!(lex);

    assert!(check_same(lexed, expected));
}

#[test]
fn test_lexing_double_symbols() {
    let source = "$!-->$->$<-$<-$>=<=@!%!";
    let expected = vec![
        token!(TokenType::RawDollar, "$"),
        token!(TokenType::Minus, "-"),
        token!(TokenType::Minus, "-"),
        token!(TokenType::Great, ">"),
        token!(TokenType::TextMathStart, "$"),
        token!(TokenType::RightArrow, "\\rightarrow "),
        token!(TokenType::TextMathEnd, "$"),
        token!(TokenType::Less, "<"),
        token!(TokenType::Minus, "-"),
        token!(TokenType::TextMathStart, "$"),
        token!(TokenType::LeftArrow, "\\leftarrow "),
        token!(TokenType::TextMathEnd, "$"),
        token!(TokenType::GreatEq, "\\geq "),
        token!(TokenType::LessEq, "\\leq "),
        token!(TokenType::At, "@"),
        token!(TokenType::LatexComment, "%"),
    ];

    let mut lex = Lexer::new(source);
    let lexed = lexing!(lex);

    assert!(check_same(lexed, expected));
}

#[test]
fn test_lexing_whitespaces() {
    let source = "\t  \t\n\r\n \t\r";
    let expected = vec![
        token!(TokenType::Tab, "\t"),
        token!(TokenType::Space, " "),
        token!(TokenType::Space, " "),
        token!(TokenType::Tab, "\t"),
        token!(TokenType::Newline, "\n"),
        token!(TokenType::Newline, "\n"),
        token!(TokenType::Space, " "),
        token!(TokenType::Tab, "\t"),
        token!(TokenType::Newline, "\n"),
    ];

    let mut lex = Lexer::new(source);
    let lexed = lexing!(lex);

    assert!(check_same(lexed, expected));
}

#[test]
fn test_lexing_comment() {
    let source = r#"
% This is a comment!
----
%* This is also a comment.
The difference is that multiple commenting is possible.
*%+"#;
    let expected = vec![
        token!(TokenType::Newline, "\n"),
        token!(TokenType::Minus, "-"),
        token!(TokenType::Minus, "-"),
        token!(TokenType::Minus, "-"),
        token!(TokenType::Minus, "-"),
        token!(TokenType::Newline, "\n"),
        token!(TokenType::Plus, "+"),
    ];

    let mut lex = Lexer::new(source);
    let lexed = lexing!(lex);

    assert!(check_same(lexed, expected));
}

#[test]
fn test_text_raw_latex() {
    let source = "%-\\TeX and \\LaTeX-%%-foo 3.14-%";
    let expected = vec![
        token!(TokenType::RawLatex, "\\TeX and \\LaTeX"),
        token!(TokenType::RawLatex, "foo 3.14"),
    ];

    let mut lex = Lexer::new(source);
    let lexed = lexing!(lex);

    assert!(check_same(lexed, expected));
}

#[test]
fn test_inline_raw_latex() {
    let source = r#"%-
\begin{center}
  the \TeX
\end{center}
-%"#;
    let expected = vec![token!(
        TokenType::RawLatex,
        r#"
\begin{center}
  the \TeX
\end{center}
"#
    )];

    let mut lex = Lexer::new(source);
    let lexed = lexing!(lex);

    assert!(check_same(lexed, expected));
}

#[test]
fn test_lexing_ascii_string() {
    let source = "This is a string!";
    let expected = vec![
        token!(TokenType::Text, "This"),
        token!(TokenType::Space, " "),
        token!(TokenType::Text, "is"),
        token!(TokenType::Space, " "),
        token!(TokenType::Text, "a"),
        token!(TokenType::Space, " "),
        token!(TokenType::Text, "string"),
        token!(TokenType::Bang, "!"),
    ];

    let mut lex = Lexer::new(source);
    let lexed = lexing!(lex);

    assert!(check_same(lexed, expected));
}

#[test]
fn test_lexing_math_character() {
    let source = "$ $\\[\\]\\[[\\]";
    let expected = vec![
        token!(TokenType::TextMathStart, "$"),
        token!(TokenType::Space, " "),
        token!(TokenType::TextMathEnd, "$"),
        token!(TokenType::InlineMathStart, "\\["),
        token!(TokenType::InlineMathEnd, "\\]"),
        token!(TokenType::InlineMathStart, "\\["),
        token!(TokenType::LeftSquareBrace, "["),
        token!(TokenType::InlineMathEnd, "\\]"),
    ];

    let mut lex = Lexer::new(source);
    let lexed = lexing!(lex);

    assert!(check_same(lexed, expected));
}

#[test]
fn test_lexing_unicode_string() {
    let source = "이것은 무엇인가?";
    let expected = vec![
        token!(TokenType::Text, "이것은"),
        token!(TokenType::Space, " "),
        token!(TokenType::Text, "무엇인가"),
        token!(TokenType::Question, "?"),
    ];

    let mut lex = Lexer::new(source);
    let lexed = lexing!(lex);

    assert!(check_same(lexed, expected));
}

#[test]
fn test_lex_number() {
    let source = "1 32 -8432 3.2 0.3 32.00 3.20";
    let expected = vec![
        token!(TokenType::Integer, "1"),
        token!(TokenType::Space, " "),
        token!(TokenType::Integer, "32"),
        token!(TokenType::Space, " "),
        token!(TokenType::Integer, "-8432"),
        token!(TokenType::Space, " "),
        token!(TokenType::Float, "3.2"),
        token!(TokenType::Space, " "),
        token!(TokenType::Float, "0.3"),
        token!(TokenType::Space, " "),
        token!(TokenType::Text, "32.00"),
        token!(TokenType::Space, " "),
        token!(TokenType::Text, "3.20"),
    ];

    let mut lex = Lexer::new(source);
    let lexed = lexing!(lex);

    assert!(check_same(lexed, expected));
}

#[test]
fn lexing_keywords() {
    let source = "docclass begenv startdoc mtxt vesmod etxt endenv";
    let expected = vec![
        token!(TokenType::Docclass, "docclass"),
        token!(TokenType::Space, " "),
        token!(TokenType::BeginEnv, "begenv"),
        token!(TokenType::Space, " "),
        token!(TokenType::StartDoc, "startdoc"),
        token!(TokenType::Space, " "),
        token!(TokenType::Mtxt, "mtxt"),
        token!(TokenType::Space, " "),
        token!(TokenType::VestiMod, "vesmod"),
        token!(TokenType::Space, " "),
        token!(TokenType::Etxt, "etxt"),
        token!(TokenType::Space, " "),
        token!(TokenType::EndEnv, "endenv"),
    ];

    let mut lex = Lexer::new(source);
    let lexed = lexing!(lex);

    assert!(check_same(lexed, expected));
}

#[test]
fn lexing_backslash() {
    let source = "\\#\\$\\)\\%";
    let expected = vec![
        token!(TokenType::Sharp, "\\#"),
        token!(TokenType::Dollar, "\\$"),
        token!(TokenType::TextMathEnd, "$"),
        token!(TokenType::Percent, "\\%"),
    ];

    let mut lex = Lexer::new(source);
    let lexed = lexing!(lex);

    assert!(check_same(lexed, expected));
}

#[test]
fn lexing_latex_functions() {
    let source = "\\foo \\bar@hand \\frac{a; b}";
    let expected = vec![
        token!(TokenType::LatexFunction, "\\foo"),
        token!(TokenType::Space, " "),
        token!(TokenType::LatexFunction, "\\bar@hand"),
        token!(TokenType::Space, " "),
        token!(TokenType::LatexFunction, "\\frac"),
        token!(TokenType::LeftBrace, "{"),
        token!(TokenType::Text, "a"),
        token!(TokenType::Semicolon, ";"),
        token!(TokenType::Space, " "),
        token!(TokenType::Text, "b"),
        token!(TokenType::RightBrace, "}"),
    ];

    let mut lex = Lexer::new(source);
    let lexed = lexing!(lex);

    assert!(check_same(lexed, expected));
}

#[test]
fn basic_vesti() {
    let source = r#"docclass coprime (tikz, korean)
ltxpkg {
    geometry (a4paper, margin = 0.4in),
    amsmath,
}

startdoc

This is a \LaTeX!
\[
    1 + 1 = \sum_{j=1}^\infty f(x),\qquad mtxt foobar etxt
\]
begenv center % [adadasdawd]
    The TeX
endenv"#;
    let expected = vec![
        token!(TokenType::Docclass, "docclass"),
        token!(TokenType::Space, " "),
        token!(TokenType::Text, "coprime"),
        token!(TokenType::Space, " "),
        token!(TokenType::LeftParen, "("),
        token!(TokenType::Text, "tikz"),
        token!(TokenType::Comma, ","),
        token!(TokenType::Space, " "),
        token!(TokenType::Text, "korean"),
        token!(TokenType::RightParen, ")"),
        token!(TokenType::Newline, "\n"),
        token!(TokenType::LatexPkg, "ltxpkg"),
        token!(TokenType::Space, " "),
        token!(TokenType::LeftBrace, "{"),
        token!(TokenType::Newline, "\n"),
        token!(TokenType::Space, " "),
        token!(TokenType::Space, " "),
        token!(TokenType::Space, " "),
        token!(TokenType::Space, " "),
        token!(TokenType::Text, "geometry"),
        token!(TokenType::Space, " "),
        token!(TokenType::LeftParen, "("),
        token!(TokenType::Text, "a4paper"),
        token!(TokenType::Comma, ","),
        token!(TokenType::Space, " "),
        token!(TokenType::Text, "margin"),
        token!(TokenType::Space, " "),
        token!(TokenType::Equal, "="),
        token!(TokenType::Space, " "),
        token!(TokenType::Float, "0.4"),
        token!(TokenType::Text, "in"),
        token!(TokenType::RightParen, ")"),
        token!(TokenType::Comma, ","),
        token!(TokenType::Newline, "\n"),
        token!(TokenType::Space, " "),
        token!(TokenType::Space, " "),
        token!(TokenType::Space, " "),
        token!(TokenType::Space, " "),
        token!(TokenType::Text, "amsmath"),
        token!(TokenType::Comma, ","),
        token!(TokenType::Newline, "\n"),
        token!(TokenType::RightBrace, "}"),
        token!(TokenType::Newline, "\n"),
        token!(TokenType::Newline, "\n"),
        token!(TokenType::StartDoc, "startdoc"),
        token!(TokenType::Newline, "\n"),
        token!(TokenType::Newline, "\n"),
        token!(TokenType::Text, "This"),
        token!(TokenType::Space, " "),
        token!(TokenType::Text, "is"),
        token!(TokenType::Space, " "),
        token!(TokenType::Text, "a"),
        token!(TokenType::Space, " "),
        token!(TokenType::LatexFunction, "\\LaTeX"),
        token!(TokenType::Bang, "!"),
        token!(TokenType::Newline, "\n"),
        token!(TokenType::InlineMathStart, "\\["),
        token!(TokenType::Newline, "\n"),
        token!(TokenType::Space, " "),
        token!(TokenType::Space, " "),
        token!(TokenType::Space, " "),
        token!(TokenType::Space, " "),
        token!(TokenType::Integer, "1"),
        token!(TokenType::Space, " "),
        token!(TokenType::Plus, "+"),
        token!(TokenType::Space, " "),
        token!(TokenType::Integer, "1"),
        token!(TokenType::Space, " "),
        token!(TokenType::Equal, "="),
        token!(TokenType::Space, " "),
        token!(TokenType::LatexFunction, "\\sum"),
        token!(TokenType::Subscript, "_"),
        token!(TokenType::LeftBrace, "{"),
        token!(TokenType::Text, "j"),
        token!(TokenType::Equal, "="),
        token!(TokenType::Integer, "1"),
        token!(TokenType::RightBrace, "}"),
        token!(TokenType::Superscript, "^"),
        token!(TokenType::LatexFunction, "\\infty"),
        token!(TokenType::Space, " "),
        token!(TokenType::Text, "f"),
        token!(TokenType::LeftParen, "("),
        token!(TokenType::Text, "x"),
        token!(TokenType::RightParen, ")"),
        token!(TokenType::Comma, ","),
        token!(TokenType::LatexFunction, "\\qquad"),
        token!(TokenType::Space, " "),
        token!(TokenType::Mtxt, "mtxt"),
        token!(TokenType::Space, " "),
        token!(TokenType::Text, "foobar"),
        token!(TokenType::Space, " "),
        token!(TokenType::Etxt, "etxt"),
        token!(TokenType::Newline, "\n"),
        token!(TokenType::InlineMathEnd, "\\]"),
        token!(TokenType::Newline, "\n"),
        token!(TokenType::BeginEnv, "begenv"),
        token!(TokenType::Space, " "),
        token!(TokenType::Text, "center"),
        token!(TokenType::Space, " "),
        token!(TokenType::Space, " "),
        token!(TokenType::Space, " "),
        token!(TokenType::Space, " "),
        token!(TokenType::Space, " "),
        token!(TokenType::Text, "The"),
        token!(TokenType::Space, " "),
        token!(TokenType::Text, "TeX"),
        token!(TokenType::Newline, "\n"),
        token!(TokenType::EndEnv, "endenv"),
    ];
    let mut lex = Lexer::new(source);
    let lexed = lexing!(lex);

    assert!(check_same(lexed, expected));
}